[snoopy.brown 김상열] 코드 리뷰 요청드립니다.

syeol-kim opened the pull request:
안녕하세요. 스누피입니다:D
지하철 노선도 미션(4단계)에 대해 코드 리뷰를 요청드립니다.

syeol-kim committed:
docs: 미션 소개 및 요구 사항 추가

syeol-kim committed:
refactor: 패키지 및 로직 분리

syeol-kim committed:
refactor: SECTION 테이블에 외래키 지정

syeol-kim committed:
refactor: 패키지 및 로직 분리/변경

가상의 구간을 임의로 만들었던 기존 방식에서, 가상의 구간을 만들지 않고 직관적으로 삽입/삭제하도록 수정함.

syeol-kim committed:
refactor: 패키지 및 로직 분리/변경

syeol-kim committed:
refactor: 예외별 HttpStatus를 지정

동일한 예외 발생 시 같은 HttpStatus를 반환하기 위해 지정함.
발생한 예외는 GlobalExceptionHandler가 처리함.

syeol-kim committed:
test: 도메인 테스트 코드 추가

syeol-kim committed:
docs: 목차 링크 수정

syeol-kim commented on src/main/java/subway/line/web/LineController.java:
```diff
@@ -0,0 +1,120 @@
+
+    @GetMapping(value = "/lines", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<List<LineResponse>> showLines() {
+        return lineService.findAllLines()
+                .stream()
+                .map(line -> {
+                    Stations stations = stationService.findStationsByIds(
+                            sectionService.findSectionsByLineId(line.getId())
+                                    .getStationIdsInDownwardOrder()
+                    );
+                    return new LineResponse(line, stations);
+                }).collect(Collectors.collectingAndThen(
+                        Collectors.toList(),
+                        ResponseEntity::ok
+                ));
+    }
```
lineService, stationService, sectionService 가 사용되어 코드가 지저분해졌습니다.
각 서비스 간에 계층구조가 없다고 판단해서 서비스를 분리했습니다.
어떻게 개선하면 좋을지 궁금합니다.

dongjoon1251 replied:
계층 구조는 없지만 현재 구현하신 구조는 domain 별로 패키지를 나누셨기 때문에 각각의 마이크로서비스를 구현한 것으로 볼 수도 있을 것 같습니다. 그래서 다른 마이크로서비스에 접근할 때 api를 호출한다고 생각해보면 하나의 service에서 다른 도메인의 api를 호출하는 service를 호출하는 것으로도 볼 수 있습니다. api 접근의 경우 순환참조가 되어도 문제가 없지만 하나의 프로젝트 내에서 호출하는 경우 순환참조로 인한 문제가 발생할 수 있으니 이점 유념하여 리팩토링해주시기 바랍니다 :)

syeol-kim replied:
순환참조가 발생하지 않도록 서비스 내에서 다른 서비스를 호출하지 않는 것 보단, 다른 서비스를 호출하면서 순환참조가 발생하지 않도록 신경쓰는 것이 더 나은 방식이군요 📝

dongjoon1251 replied:
음 항상 더 나은 방식이라기보다는 비즈니스 로직을 최대한 service layer로 분리해보면, 그다음으로 도메인쪽으로 로직을 옮겨볼 수 있는 힌트가 될 수 있습니다 :)

dongjoon1251 commented on README.md:
```diff
@@ -13,42 +13,263 @@
 카카오 신입사원 교육 - 스프링 과정 실습을 위한 지하철 노선도 애플리케이션
 
-<br>
+## 목차
```
README를 잘 작성해주셨네요 💯

dongjoon1251 commented on src/main/java/subway/common/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,58 @@
+
+import java.sql.SQLException;
+
+@RestControllerAdvice
```
Advice를 잘 적용해주셨네요 👍 
다만 Exception 처리 로직 중복이 많아보입니다. 
enum으로 특정 에러 클래스에 대해서 다른 Response를 전달하도록 처리하면 어떨까요?
ExceptionHandler는 공통의 상위 Exception 클래스를 통해서 호출되도록 리팩토링해보면 어떨까요?

syeol-kim replied:
enum에서 Exception 마다의 HttpStatus를 반환하고, 이를 호출하여 Response를 생성할 수 있도록 반영했습니다. 또한, 공통의 상위 Exception 클래스를 핸들링하여 코드의 중복을 줄일 수 있음을 경험했습니다. 📝

만약 커스텀 예외 클래스에 하위 클래스를 생성한다면, ExceptionHandler에서 커스텀 예외 클래스를 핸들링하도록 구현하고 HttpStatus를 enum에서 관리하지 않고 ExceptionHandler에서 선언하는게 더 직관적일 꺼 같습니다. (이 프로젝트에서는 ExceptionHandler에 하나의 핸들링 메서드를 갖고 있어 enum만 보면 맵핑되는 HttpStatus를 파악할 수 있지만, 상위 커스텀 예외 클래스가 많아지고 각각의 핸들링 메서드를 만든다면 하위 클래스의 HttpStatus를 보기 위해 ExceptionHandler와 enum를 둘 다 확인해야 하므로 직관성이 떨어진다고 생각합니다.)

dongjoon1251 commented on src/main/java/subway/common/error/ErrorStatus.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.http.HttpStatus;
+
+public enum ErrorStatus {
+    // Common
```
불필요한 주석은 제거해주세요:)

dongjoon1251 commented on src/main/java/subway/line/dao/LineDao.java:
```diff
@@ -0,0 +1,18 @@
+import subway.line.vo.Lines;
+
+import java.util.Optional;
+
```
DAO interface 분리 👍

dongjoon1251 commented on src/main/java/subway/line/dao/LineDaoImpl.java:
```diff
@@ -0,0 +1,85 @@
+import java.util.Optional;
+
+@Repository
+@Transactional(propagation = Propagation.MANDATORY)
```
모든 dao에는 propagation = Propagation.MANDATORY 옵션이 추가되야 할까요? 고민해주세요 :)

syeol-kim replied:
transactional 어노테이션에 대해 공부하던 중 [stackoverflow](https://stackoverflow.com/a/7239615)을 읽고, service에 transactional 어노테이션을 주고  dao 에서는 MANDATORY 옵션을 주면 service에 transactional을 붙이지 않는 등 오류를 감지할 수 있으니, 반드시 MANDATORY 옵션을 사용해야하는 건 아니지만, 제가 채택한 방식에서 유용하다고 생각했습니다. 🤔

syeol-kim replied:
transactional 어노테이션에 대해 공부하던 중 [stackoverflow](https://stackoverflow.com/a/7239615)을 읽고, service에 transactional 어노테이션을 주고  dao 에서는 MANDATORY 옵션을 주면 service에 transactional을 붙이지 않는 등 오류를 감지할 수 있으니, 반드시 MANDATORY 옵션을 사용해야하는 건 아니지만, 제가 채택한 방식에서 유용하다고 생각했습니다. 🤔

dongjoon1251 commented on src/main/java/subway/line/dao/LineDaoImpl.java:
```diff
@@ -0,0 +1,85 @@
+    public Line insert(Line line) {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbcTemplate.update(con -> {
+            PreparedStatement ps = con.prepareStatement(
```
indent가 두번 되었네요. 메소드로 분리하면 어떨까요?

syeol-kim replied:
반영했습니다 🤗

syeol-kim replied:
반영했습니다 🤗

dongjoon1251 commented on src/main/java/subway/line/dao/LineDaoImpl.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public Optional<Line> findLineById(Long id) {
+        try {
```
여기도 인덴트가 2번 발생했네요. try-catch 구문을 Optional 로 감싸도록 리팩토링해보면 어떨까요?

syeol-kim replied:
try-catch 구문을 Optionial로 감싸면 indent가 두번 되지않나요? 이해가 되지 않습니다. 🤔
(주석처리한 부분까지) 두 가지 방식으로 indent를 줄여봤습니다.

syeol-kim replied:
try-catch 구문을 Optionial로 감싸면 indent가 두번 되지않나요? 이해가 되지 않습니다. 🤔
의도하신 방식은 아닌듯 하지만 indent를 줄여봤습니다.

dongjoon1251 replied:
try-catch 구문을 별도 메소드로 분리하고 해당 메소드를 Optional로 감싸는 방식을 제안드린거였는데 전달이 잘 안되었네요:)
지금과 같은 방식으로 구현해주셔도 무방합니다.

dongjoon1251 commented on src/main/java/subway/line/dto/LineRequest.java:
```diff
@@ -37,4 +40,16 @@ public Long getDownStationId() {
         return distance;
     }
+
+    public Line toLine() {
```
dto에서 필요한 필드를 가져와서 생성하는 것이 아니라 생성자를 호출하는 로직을 넣으면, 해당 메소드를 호출할 때마다 객체가 새로 생성되게 됩니다. 더 좋은 방법이 없을지 고민해주시기 바랍니다.

syeol-kim replied:
말씀해주신대로 dto가 생성자를 호출하면 매번 새로운 객체를 생성하게 됩니다. 저는 [mapstruct](https://mapstruct.org/)를 보고 이 방식이 일반적이라고 생각했습니다. (request 객체로 부터 entity 생성을 위한 dto를 만들고, dao에서 insert 파라미터로 dto를 받고 entity를 반환하도록 구현하는 방식)

새로운 객체를 생성하지 않고 값을 전달하려면 request 객체를 그대로 전달하거나, 각 필드값을 getter로 꺼내서 전달하는 방식을 말씀하시는 걸까요? 🤔  request 객체를 서비스에 파라미터로 사용할 경우 재사용성이 떨어지게되므로, 필드값을 꺼내어 전달하는 방식으로 리펙토링했습니다.

dongjoon1251 replied:
이해하신 방식이 맞습니다. 잘 적용해주셨습니다 👍

dongjoon1251 commented on src/main/java/subway/line/service/LineService.java:
```diff
@@ -0,0 +1,16 @@
+import subway.line.vo.Line;
+import subway.line.vo.Lines;
+
+public interface LineService {
```
service 인터페이스 분리 👍 
단, 구현체가 하나만 있을 확률이 높을때에는 interface를 사용하지 않기도 합니다.

syeol-kim replied:
처음 스프링을 배울때 썼던 것이 습관이 되어버렸네요 ㅎㅎ;

dongjoon1251 commented on src/main/java/subway/line/service/LineServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Override
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
```
여러명이 수정할 경우를 고려하여 Isolation.REPEATABLE_READ 를 사용하신 것으로 보입니다. 이러한 문제를 해결하기 위한 다른 방법에 대해서도 공부해보시면 좋을 것 같습니다. 
힌트)  lock

syeol-kim replied:
[Optimistic/Pessimistic Lock](https://effectivesquid.tistory.com/entry/Optimistic-Lock과-Pessimistic-Lock)을 참고했을때, 여러명이 수정할 경우에는 PESSIMISTIC_WRITE를 사용할 수 있겠네요! 📝

dongjoon1251 commented on src/main/java/subway/line/service/LineServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+            throw new NotExistEntityException("존재하지 않는 지하철 노선입니다.");
+        }
+
+        if (isNotUpdated(lineDao.delete(id))) {
```
isolation = Isolation.REPEATABLE_READ 을 지정하고, 존재 유무도 확인했는데 0이 리턴될 수 있을까요? 한번 고민해주세요 :)

syeol-kim replied:
0을 리턴하는 경우는 없을듯 합니다. 방어적 프로그래밍을 하다보니 더블 체크가 되버렸네요. 😅

dongjoon1251 commented on src/main/java/subway/line/vo/Line.java:
```diff
@@ -1,9 +1,9 @@
-package subway.line;
+package subway.line.vo;
```
value object와 entity의 차이점에 대해서 공부해보시고 적합한 패키지명을 적용해주세요 :)

syeol-kim replied:
vo와 entity를 구분하지 않고 사용해왔던거 같습니다. 정의와 차이점을 간단히 정리했습니다. 📝
* vo는 값을 담는 객체이며, 객체간의 비교를 위해 사용합니다.
* entity는 db의 테이블에 1:1 맵핑되며, 해당 테이블내에 존재하는 컬럼만을 필드로 갖습니다.
* vo는 불변성을 entity는 가변성을 갖습니다.

dongjoon1251 replied:
가장 중요한 차이점은 entity는 ID를 가져서 각각의 객체를 식별 가능하다는 점입니다 :)

dongjoon1251 commented on src/main/java/subway/line/web/LineController.java:
```diff
@@ -0,0 +1,104 @@
+
+    @PostMapping("/lines")
+    public ResponseEntity<LineResponse> createLine(@RequestBody LineRequest lineRequest) {
+        Line line = lineService.create(
```
여러 개의 서비스를 호출하여 조합하는 것도 비즈니스 로직으로 볼 수 있습니다. controller에서는 최대한 하나의 service만 접근하도록 리팩토링해주세요 :) service에서 service를 호출해도 무방합니다. 단, 순환 참조가 되지 않도록 주의해주세요.

syeol-kim replied:
말씀해주신대로 서비스 내부에서 서비스를 호출하도록 반영했습니다 🤗

dongjoon1251 commented on src/main/java/subway/section/service/SectionServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+
+    @Override
+    public void connect(Section section) {
+        Sections sections = sectionDao.findSectionsByLineId(section.getLineId());
```
해당 로직을 최대한 sections 로 이동하도록 고민해보면 어떨까요?
추가할 section을 sections로 전달하고, 추가하거나 삭제해야할 section을 가져오는 메소드를 만들면 어떨까요?
service에서는 dao 호출을 통해서 CRUD 작업만 주로 다룬다고 생각하시고 리팩토링해주시기 바랍니다 :)

syeol-kim replied:
리펙토링 과정을 거치면서, 기존에는 단순히 구간들을 Sections로 정의했지만, 반드시 연결되어 있는 구간들은 LineSections로, 연결 여부에 관계없는 구간들은 Sections로 분리시켰습니다.
분리시킨 후 다시 코드를 봤을 때, dao와의 커뮤니케이션을 보아 로직을 이동시키기 어렵다고 판단했습니다. 😭

dongjoon1251 commented on src/main/java/subway/section/service/SectionServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+    @Override
+    public void deleteByLineId(Long lineId) {
+        // 지하철 노선을 삭제할 때는 삭제 가능 여부를 검사하지 않는다 (강제 삭제)
+        sectionDao.findSectionsByLineId(lineId)
```
list를 삭제하는 method를 dao에 추가해보면 어떨까요?

syeol-kim replied:
> 삭제되거나 변경되기 전에 해당 도메인이 존재하는지 확인하는 부분을 추가해보면 어떨까요? 그러면 삭제나 업데이트 로직이 왜 **실패**했는지 **원인파악**이 용이할 수 있어 보입니다.

제노스가 이전 단계에서 주신 피드백을 반영하면서 수정/삭제 쿼리전에 db 조회 쿼리를 추가했고 update 성공/실패를 service가 로직에 반영하도록 했습니다. 이를 list에 적용하면, 수정/삭제할 때는 (데이터베이스에 자주 접근하더라도) 각 원소의 성공/실패를 확인해야 한다고 생각했습니다. 
list로 수정/삭제하는 쿼리는 성공/실패를 service에서 로직에 반영할 필요가 없음(성공/실패가 로직에 영향을 끼치지 않음)을 가정하고 사용한다고 이해하면 될까요? 🤔

dongjoon1251 replied:
말씀해주신 것처럼 validation을 위해서 개별적으로 커밋하는 것도 맞는 방법입니다 :)
쿼리 한번에 전체 리스트를 추가하는 부분에 대해서 고민해보실 수 있을 것 같아서 코멘트 드렸었습니다.

dongjoon1251 commented on src/main/java/subway/section/service/SectionServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+    @Override
+    public void deleteByLineIdAndStationId(Long lineId, Long stationId) {
+        // 지하철 역을 삭제할 때는 삭제 가능 여부를 검사한다
+        Sections sections = sectionDao.findSectionsByLineId(lineId);
```
이부분도 sections에서 station을 삭제하는 메소드, 삭제하면서 변경되야할 section을 가져오는 메소드로 분리해보면 어떨까요?

syeol-kim replied:
리펙토링 과정을 거치면서, 기존에는 단순히 구간들을 Sections로 정의했지만, 반드시 연결되어 있는 구간들은 LineSections로, 연결 여부에 관계없는 구간들은 Sections로 분리시켰습니다.
분리시킨 후 다시 코드를 봤을 때, dao와의 커뮤니케이션을 보아 로직을 이동시키기 어렵다고 판단했습니다. 😭

syeol-kim replied:
리펙토링 과정을 거치면서, 기존에는 단순히 구간들을 Sections로 정의했지만, 반드시 연결되어 있는 구간들은 LineSections로, 연결 여부에 관계없는 구간들은 Sections로 분리시켰습니다. 분리시킨 후 다시 코드를 봤을 때, dao와의 커뮤니케이션을 보아 로직을 이동시키기 어렵다고 판단했습니다. 😭

dongjoon1251 commented on src/main/java/subway/section/vo/Section.java:
```diff
@@ -0,0 +1,133 @@
+    private Long id;
+
+    public Section(long lineId, long upStationId, long downStationId, int distance) {
+        validate(upStationId, downStationId, distance);
```
가독성을 위해서 메소드 내부에서 사용된 외부 메소드에 대한 정의는 바로 하단에 정의해주세요:)

syeol-kim replied:
메서드 정의 위치에 좀 더 주의를 기울이겠습니다. 🔥

dongjoon1251 commented on src/main/java/subway/section/vo/Sections.java:
```diff
@@ -0,0 +1,166 @@
+        List<Long> endStationIds = stationIds.stream()
+                .distinct()
+                .filter(stationId -> Collections.frequency(stationIds, stationId) == 1)
+                .collect(Collectors.toList());
```
sections를 정렬할때 사용하는 코드로 확인됩니다. 현재는 O(n*n)인 것 같습니다. O(n)으로 줄일수 있는 방법에 대해서 고민해보시면 어떨까요? 성능 개선이 미션의 목적은 아니지만 간단한 알고리즘은 가독성도 높여줄 수 있습니다.

syeol-kim replied:
성능에 좀 더 주의를 기울이겠습니다. 🔥

dongjoon1251 commented on src/main/java/subway/section/vo/Sections.java:
```diff
@@ -0,0 +1,166 @@
+
+    private boolean isNotUnited(List<Section> sections) {
+        assert !isEmpty(sections);
+        Map<Long, Section> cache = sections.stream()
```
sections에서 조회하는 경우가 많은데 해당 cache를 생성시점에 필드로 가지면 어떨까요?

syeol-kim replied:
처음에는 cache를 멤버 변수로 만들고 생성시에 초기화도록 작성했습니다.  
  이어지는 리펙터링 과정에서 __일급 컬렉션은 다른 멤버 변수를 갖지 않는다__ 에 따라 cache를 임시로 생성하도록 수정했습니다.
  > 규칙 8: 일급 콜렉션 사용
  이 규칙의 적용은 간단하다.  
  콜렉션을 포함한 클래스는 반드시 다른 멤버 변수가 없어야 한다.  
  각 콜렉션은 그 자체로 포장돼 있으므로 이제 콜렉션과 관련된 동작은 근거지가 마련된셈이다.  
  필터가 이 새 클래스의 일부가 됨을 알 수 있다.  
  필터는 또한 스스로 함수 객체가 될 수 있다.  
  또한 새 클래스는 두 그룹을 같이 묶는다든가 그룹의 각 원소에 규칙을 적용하는 등의 동작을 처리할 수 있다.  
  이는 인스턴스 변수에 대한 규칙의 확실한 확장이지만 그 자체를 위해서도 중요하다.  
  콜렉션은 실로 매우 유용한 원시 타입이다.  
  많은 동작이 있지만 후임 프로그래머나 유지보수 담당자에 의미적 의도나 단초는 거의 없다. - 소트웍스 앤솔로지 객체지향 생활체조편

dongjoon1251 commented on src/main/java/subway/station/service/StationServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+    @Transactional(readOnly = true)
+    public Stations findStationsByIds(List<Long> ids) {
+        List<Station> stations = new ArrayList<>();
+        for (Long id : ids) {
```
id list를 받아서 List<Station>을 조회하는 하나의 쿼리를 사용하는 dao method를 만들면 어떨까요?

syeol-kim replied:
반영했습니다 🤗

dongjoon1251 commented on src/main/java/subway/station/vo/Stations.java:
```diff
@@ -0,0 +1,17 @@
+
+public class Stations {
+    private final List<Station> stations;
+
```
stations내에 name 중복이 없는지 체크하는 로직이 추가되는게 어떨까요?

syeol-kim replied:
반영했습니다 🤗

dongjoon1251 reviewed (changes requested):
피드백과 함께 리팩토링을 위해 많은 부분들을 고민해주시고 반영해주셨네요 👍 
몇가지 코멘트를 드렸으니 확인 부탁드립니다 :)
그리고 controller, service부분에 domain로직이 아직 많이 포함되어 있습니다. 최대한 더 domain 로직에 포함되도록 리팩토링해주시고, 도메인별로 테스트코드를 다양한 케이스에 대해서 작성해주세요.

syeol-kim committed:
chore: 패키지명 수정

vo -> entity

syeol-kim committed:
refactor: 예외 핸들링 메서드 중복 코드 제거

syeol-kim committed:
style: 메서드 정의 위치 변경

가독성을 높이기 위해 메서드 위치를 변경함

syeol-kim committed:
feat: 유효성 검사 기능 추가

syeol-kim committed:
refactor: Sections 개념 재정의 및 분리

before:
  List<Section>을 Sections로 정의
after:
  연결성에 무관한 List<Section>을 Sections
  순서대로 연결성을 갖는 List<Section>을 LineSections로 정의함.

syeol-kim committed:
refactor: dto To entity 제거

before:
  dto 에서 새로운 객체(entity)를 만들어서 전달
after:
  dto에서 getter로 필드 값을 전달

indent를 줄이기 위해 코드 수정
dao에서 리스트로 수정/삭제하는 기능 추가

syeol-kim committed:
refactor: 미사용 코드 제거

syeol-kim committed:
test: 도메인 테스트 추가

syeol-kim committed:
docs: 피드백 파일 제외

syeol-kim commented on src/main/java/subway/line/dao/LineDaoImpl.java:
```diff
@@ -36,37 +38,49 @@ public LineDaoImpl(JdbcTemplate jdbcTemplate) {
+//        } catch (EmptyResultDataAccessException e) {
+//            return Optional.empty();
+//        }
+//        return Optional.ofNullable(line);
```
try-catch 구문을 Optionial로 감싸면 indent가 두번 되지않나요? 이해가 되지 않습니다. 🤔
(주석처리한 부분까지) 두 가지 방식으로 indent를 줄여봤습니다.

syeol-kim commented on src/main/java/subway/line/dao/LineDaoImpl.java:
```diff
@@ -36,37 +38,49 @@ public LineDaoImpl(JdbcTemplate jdbcTemplate) {
     }
 
     @Override
     public Optional<Line> findLineById(Long id) {
```
try-catch 구문을 Optionial로 감싸면 indent가 두번 되지않나요? 이해가 되지 않습니다. 🤔
 (주석처리한 부분까지) 두 가지 방식으로 indent를 줄여봤습니다.

syeol-kim commented on src/main/java/subway/line/web/LineController.java:
```diff
@@ -32,52 +31,35 @@ public LineController(LineService lineService, StationService stationService, Se
-                        line.getId()
-                )
+                lineRequest.getName(),
+                lineRequest.getColor()
```
말씀해주신대로 dto가 생성자를 호출하면 매번 새로운 객체를 생성하게 됩니다. 저는 [mapstruct](https://mapstruct.org/)를 보고 이 방식이 일반적이라고 생각했습니다. (request 객체로 부터 entity 생성을 위한 dto를 만들고, dao에서 insert 파라미터로 dto를 받고 entity를 반환하도록 구현하는 방식)

새로운 객체를 생성하지 않고 값을 전달하려면 request 객체를 그대로 전달하거나, 각 필드값을 getter로 꺼내서 전달하는 방식을 말씀하시는 걸까요? 🤔   request 객체를 서비스에 파라미터로 사용할 경우 재사용성이 떨어진다고 생각하여 필드값을 꺼내어 전달하는 방식으로 리펙토링했습니다.

syeol-kim commented on src/main/java/subway/line/web/LineController.java:
```diff
@@ -32,52 +31,35 @@ public LineController(LineService lineService, StationService stationService, Se
-                        Collectors.toList(),
-                        ResponseEntity::ok
-                ));
+        return ResponseEntity.ok(lineService.getAllLinesWithStations());
```
말씀해주신대로 서비스 내부에서 서비스를 호출하도록 반영했습니다  🤗

syeol-kim commented on src/main/java/subway/section/dao/SectionDaoImpl.java:
```diff
@@ -39,46 +43,52 @@ public SectionDaoImpl(JdbcTemplate jdbcTemplate) {
-            ps.setInt(4, section.getDistance());
-            return ps;
-        }, keyHolder);
+        PreparedStatementCreator psc = generatePreparedStatementCreator(lineId, upStationId, downStationId, distance);
+        jdbcTemplate.update(psc, keyHolder);
```
다음 미션을 진행하면서 알게된 사실은 SimpleJdbcInsert을 사용하면 더 간단해진다는 점.. 😅

syeol-kim commented on src/main/java/subway/station/entity/Stations.java:
```diff
@@ -35,4 +39,22 @@ private boolean hasDuplicateName(List<Station> stations) {
+        return stations.size() == size;
+    }
+
+    public Stations sortByIds(List<Long> ids) {
+        Map<Long, Station> identityMap = getIdentityMap();
```
기존에는 순서대로 db에서 꺼내오는 방식을 썼지만, db에서 리스트로 읽어오고 로직으로 정렬하도록 수정했습니다. 🤗

syeol-kim reviewed (commented):
피드백 반영해서 리펙토링 했습니다 :D

dongjoon1251 commented on src/main/java/subway/common/error/ErrorStatus.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public static ErrorStatus of(Exception e) {
+        if (e instanceof DataAccessException) {
```
enum으로 status  매핑을 잘 적용해주셨네요 👍 
instanceof로 Exception이 늘어날때마다 코드가 늘어날것으로 예상됩니다.
values, stream 등을 사용하여 리팩토링해보면 어떨까요?
각 enum이 대응되는 HttpStatus와 함께 대응되는 상위 예외클래스를 필드로 가져도 좋을 것 같습니다.

syeol-kim replied:
반영했습니다 🤗  코드가 한결 깔끔해졌어요!

dongjoon1251 commented on src/main/java/subway/section/entity/Sections.java:
```diff
@@ -0,0 +1,88 @@
+        return sections.stream()
+                .filter(section -> endStationIds.contains(section.getUpStationId()))
+                .findAny()
+                .orElse(null);
```
null이 리턴되야 하면 외부에서 Optional.ofNullable로 사용하게 유도하거나 에러를 발생시키면 어떨까요?
null 사용은 최대한 지양해야 합니다.

syeol-kim replied:
내부에서 에러를 던지도록 반영했습니다 🤗

dongjoon1251 commented on src/main/java/subway/section/entity/Sections.java:
```diff
@@ -0,0 +1,88 @@
+
+    private Section getUpEndSection() {
+        List<Long> endStationIds = getEndStationIds();
+//        return sections.stream()
```
사용하지 않는 주석은 제거해주세요:)

syeol-kim replied:
'질문해야지!' 하고 주석으로 남겨뒀었는데 잊어버렸네요 😅
주석부분 51라인에 'list가 항상 1개의 원소를 갖는다'고 할때, assert를 달고 return list.get(0)를 하는 것이 좋을지 아니면 findAny()를 써서 반환하는 것이 좋은지 궁금합니다. 🤔 assert는 어떤 의도로 코딩했는지를 보다 명확히 전달할 수 있고, findAny는 같은 기능을 수행하면서도 보다 간결하고 추측할 수 있는 정도라고 생각합니다.

dongjoon1251 replied:
이부분은 정답은 없지만 개인적으로 findAny를 썼을 때 의미가 더 잘 전달되는 것 같습니다 :)

dongjoon1251 reviewed (changes requested):
피드백을 잘 반영해주셨네요 👍 
간단한 코멘트 몇가지를 추가로 드렸습니다. 확인 부탁드려요!

syeol-kim committed:
refactor: null을 반환하는 대신 예외를 던지도록 변경

syeol-kim committed:
refactor: 주석 제거

syeol-kim committed:
refactor: enum 필드에 예외 클래스 추가

코드의 간결함 및 가독성을 위한 필드 추가

dongjoon1251 reviewed (approved):
피드백 잘 반영해주셨네요:)
질문주신 부분은 댓글로 답변 달았습니다.
고생하셨습니다. 이번 미션은 여기서 마무리 짓도록 하겠습니다.
